src/optimizer/plan_generator.cpp

```diff
@@ -177,6 +177,88 @@ void PlanGenerator::Visit(const PhysicalDistinct *) {
+  for (auto &expr : op->left_keys) {
+    PELOTON_ASSERT(children_expr_map_[0].find(expr.get()) !=
+              children_expr_map_[0].end());
+    left_keys.push_back(children_expr_map_[0][expr.get()]);
+  }
+  for (auto &expr : op->right_keys) {
+    PELOTON_ASSERT(children_expr_map_[1].find(expr.get()) !=
+              children_expr_map_[1].end());
+    right_keys.emplace_back(children_expr_map_[1][expr.get()]);
+  }
+
+  unique_ptr<planner::AbstractPlan> join_plan =
+      unique_ptr<planner::AbstractPlan>(new planner::NestedLoopJoinPlan(
+          op->type, move(join_predicate), move(proj_info), proj_schema,
+          left_keys, right_keys));
+
+  join_plan->AddChild(move(children_plans_[0]));
+  join_plan->AddChild(move(children_plans_[1]));
+  output_plan_ = move(join_plan);
+}
+
+void PlanGenerator::Visit(const PhysicalHashJoin *op) {
+  std::unique_ptr<const planner::ProjectInfo> proj_info;
+  std::shared_ptr<const catalog::Schema> proj_schema;
+  GenerateProjectionForJoin(proj_info, proj_schema);
+
+  auto join_predicate =
+      expression::ExpressionUtil::JoinAnnotatedExprs(op->join_predicates);
+  expression::ExpressionUtil::EvaluateExpression(children_expr_map_,
+                                                 join_predicate.get());
+  expression::ExpressionUtil::ConvertToTvExpr(join_predicate.get(),
+                                              children_expr_map_);
+
+  vector<unique_ptr<const expression::AbstractExpression>> left_keys;
+  vector<unique_ptr<const expression::AbstractExpression>> right_keys;
+  vector<ExprMap> l_child_map{move(children_expr_map_[0])};
+  vector<ExprMap> r_child_map{move(children_expr_map_[1])};
+  for (auto &expr : op->left_keys) {
+    auto left_key = expr->Copy();
+    expression::ExpressionUtil::EvaluateExpression(l_child_map, left_key);
+    left_keys.emplace_back(left_key);
+  }
+  for (auto &expr : op->right_keys) {
+    auto right_key = expr->Copy();
+    expression::ExpressionUtil::EvaluateExpression(r_child_map, right_key);
+    right_keys.emplace_back(right_key);
+  }
+  // Evaluate Expr for hash plan
+  vector<unique_ptr<const expression::AbstractExpression>> hash_keys;
+  for (auto &expr : op->right_keys) {
+    auto hash_key = expr->Copy();
+    expression::ExpressionUtil::EvaluateExpression(r_child_map, hash_key);
+    hash_keys.emplace_back(hash_key);
+  }
+
+  unique_ptr<planner::HashPlan> hash_plan(new planner::HashPlan(hash_keys));
+  hash_plan->AddChild(move(children_plans_[1]));
+
+  auto join_plan = unique_ptr<planner::AbstractPlan>(new planner::HashJoinPlan(
+      op->type, move(join_predicate), move(proj_info), proj_schema,
+      left_keys, right_keys, settings::SettingsManager::GetBool(
+                                 settings::SettingId::hash_join_bloom_filter)));
+
+  join_plan->AddChild(move(children_plans_[0]));
+  join_plan->AddChild(move(hash_plan));
+  output_plan_ = move(join_plan);
+}
+
 void PlanGenerator::Visit(const PhysicalInnerNLJoin *op) {
```

Comment:
Delete the methods for `PhysicalInnerNLJoin` and `PhysicalInnerHashJoin` if we are consolidating Inner+Outer joins

Reply:
Agreed.